doc/news/_preparation_next_release.md

```diff
@@ -73,6 +73,46 @@ Take a look at the [new docs](../dev/mountpoints.md), if you need to know detail
+This means noticeably reduced memory usage for cases where keys and keysets are copied and/or duplicated!
+
+We run some very promising benchmarks, each were performed with 400,000 keys.
+All benchmarks were executed using `valgrind --tool=massif --time-unit=B --max-snapshots=200 --threshold=0.1`.
+
+| Benchmark      | Current Implementation | Copy-on-Write | Size Reduction | Remarks                    |
+| :------------- | ---------------------: | ------------: | -------------: | :------------------------- |
+| `createkeys.c` |                5.3 MiB |       6.5 MiB |          -22 % |                            |
+| `deepdup.c`    |               10.5 MiB |       8.2 MiB |           22 % |                            |
+| `large.c`      |               18.9 MiB |      15.3 MiB |           19 % |                            |
+| `kdb.c`        |               23.5 MiB |      17.8 MiB |           24 % |                            |
+| `kdbget.c`     |               11.0 MiB |       8.8 MiB |           20 % |                            |
+| `kdbmodify.c`  |               11.0 MiB |       8.8 MiB |           20 % | Same results as `kdbget.c` |
+
+First, it should be noted that a single key, without counting payload, is about 50% larger with the copy-on-write implementation.
+This explains why the `createkeys.c` benchmark yields a negative reduction result.
+This benchmark only allocates keys, so not much improvement can be expected there.
+Still, as other stuff also uses heap memory, the overall memory consumption only increased by 22%, which is far less than 50%.
+
+All other benchmarks saw meaningful reductions of heap memory used.
+One interesting observation is that `kdbget.c` and `kdbmodify.c` used exactly the same memory.
+This can most likely be explained by internal caching within the memory allocator of `glibc`.
+
+We also performed runtime tests on the same benchmarks using `perf stat --repeat 13`.
+
+| Benchmark       | Current Implementation | Deviation | Copy-on-Write | Deviation | Runtime Increase |
+| :-------------- | ---------------------: | --------: | ------------: | --------: | ---------------: |
+| `createkeys.c`  |             0.209572 s |    0.36 % |     0.21987 s |    0.77 % |            4.9 % |
+| `deepdup.c`     |              0.23025 s |    0.47 % |    0.231804 s |    0.32 % |            0.6 % |
+| `large.c`       |              1.14038 s |    0.21 % |     1.14837 s |    0.21 % |            0.7 % |
+| `kdb.c`         |               1.9270 s |    2.63 % |     1.93354 s |    0.17 % |            0.3 % |
+| `kdbget.c`      |             0.145663 s |    0.17 % |     0.15763 s |    0.70 % |            8.2 % |
+| `kdbmodify.c`   |             0.146506 s |    0.19 % |    0.156347 s |    0.15 % |            6.7 % |
+
+Overall, the runtime performance hit is less than 10%.
+The more a program does, the less the additional overhead of the copy-on-write algorithms matter.
```

Comment:
I guess these benchmarks mainly serve to show that while memory usage has gone done quick a bit, the runtime is mostly unaffected.

One important thing I would also highlight is that `keyDup` has become quite a bit faster. You can see this by comparing the difference between `createkeys.c` and `deepdup.c`, since `deepdup.c` is just `createkeys.c` plus a round of `keyDup`s. In "Current Implementation" the difference is ~21ms in COW it is only ~12ms. And that is with the naive `deepdup` approach that doesn't pre-allocate the right size of `KeySet`.

Finally, I would call it "Old Implementation" instead of "Current implementation". At the time these release notes are released, COW will be the current implementation.

Reply:
> I guess these benchmarks mainly serve to show that while memory usage has gone done quick a bit, the runtime is mostly unaffected.

Yes

> One important thing I would also highlight is that keyDup has become quite a bit faster. 

Nice catch!

> Finally, I would call it "Old Implementation" instead of "Current implementation".

Good point.